### ISSUE TITLE
Using Descriptor Array in shader caused vkCreateGraphicsPipelines() failed on pre-A9 system (but working on A9 and later)

### ISSUE THREAD

**Issue by HongkunWang** [KhronosGroup/MoltenVK#446]
Hi,
I start using descriptor array in fragment shader to use multiple textures when rendering objects within a single command buffer. This solution is quite handy and is recommended by few Vulkan specialists.
So the fragment shader looks like:

```
layout(set = 1, binding = 0) uniform texture2D tex[20];   //Array holds up to 20 images
layout(push_constant) uniform per_frame_PC 
{
	uint imageId;		//Used to choose the texture for objects
};
...
void main() 
{
	outColor = texture( sampler2D(tex[imageId], separateSampler), uv);
}

```

When I test the app on few iPhones and iPads, it works well on A9 powered devices but fails on pre-A9 devices (actually iPhone 6 plus with A8 processor). The vkCreateGraphicsPipelines() function fails with initialization error (-3). 

I would like to know whether the descriptor array is supposed not to be supported by MoltenVK on all pre-A9 processor or this is a bug to be fixed. 
By the way, I noticed that numbers of Vulkan featured are not implemented on pre-A9 processor, is there a detailed list shows all those unsupported features on pre-A9 processor?

**Comment by billhollings** [comment]
Are you seeing any MoltenVK errors logged before that initialization error?

Arrays of textures should be supported on A8 processors as long as you are running at least iOS 10.

You can get an idea of what the version of Metal and device processor will support by calling the `vkGetPhysicalDeviceMetalFeaturesMVK()` function.

Khronos has a Vulkan Portability initiative which is intending to create a standard Vulkan extension to allow the app to query what Vulkan features are not supported by any particular Vulkan implementation...and MoltenVK will support that Khronos standard when it is released.

**Comment by HongkunWang** [comment]
Hi Bill,
There is not any MoltenVK error message came out before I saw VK_ERROR_INITIALIZATION_FAILED raised up from VkCreateGraphicsPipeline(). 
Here are more info I collected from my test:
The maximum uniform array size limits on each platform (tested):
		Window 10:		max array size > 65535 (like almost unlimited)
 		macOS 10.14:		max array size = 128
 		iOS 12:  			max array size = 31 (needs A9 or higher processor)
 		Android 9: 		max array size = 79
By the way, on array size limit on iOS (31) is quite inconvenient for an average game if I plan to use a single uniform buffer array to host all textures for a game scene. Is this limited by Metal API on iOS or there is other reason?
Today I tested it on iOS with A8 processor, not matter how small number I use to claim the uniform array, it always failed to create the pipeline.  My testing app is a tiny project modified from a Vulkan tutorial (a spinning cube), I attached the project here, do not know if you can run it at your side to see what is going on for A8 processor. The project can be built for all the platforms I have mentioned.
[11a_Cube.zip](https://github.com/KhronosGroup/MoltenVK/files/2748232/11a_Cube.zip)

**Comment by billhollings** [comment]
Yes...Metal supports a maximum of 31 (iOS) or 128 (macOS) textures , 31 buffers, and 16 samplers in any graphic or compute shader.

That restriction has not changed over time...and should not depend on the GPU (A9 vs A8).

Thanks for the demo app that shows the issue. I'll see if I can dig up an A8 processor device and give it a whirl.

**Comment by cdavis5e** [comment]
This is a bug in Apple's Metal compiler. You should file a radar with Apple and supply the shader that broke the compiler.

**Comment by HongkunWang** [comment]
Update: as soon as I stop using texture array in fragment shader, pipeline compilation succeeded on A8 processor.
`layout(set = 2, binding = 0) uniform texture2D tex[TEXTURE_ARRAY_SIZE]; <<<==== array`
However, I did not test regular uniform buffer array on A8.
If this is a bug in Metal compiler, then the conclusion is that we can not use texture array (or even buffer array?) with MoltenVK on A8 processor, there is no such problem on A9 and later processors.

**Comment by cdavis5e** [comment]
Hmm... I've been looking at the [Metal feature tables](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf). According to this, the A8 actually doesn't support arrays of textures. That would explain why the compiler crashed. Also, this is why the `shaderSampledImageArrayDynamicIndexing` and `shaderStorageImageArrayDynamicIndexing` features aren't turned on unless you have an A9 or better.

**Comment by HongkunWang** [comment]
Hi Davis,
Thanks for your info, it means all system powered by  PowerVR GX6450 GPU may not support arrays of textures. 
Hope MoltenVK validation layer can provide some error message instead of quietly crash the application.

**Comment by billhollings** [comment]
This is fixed in PR #463. An error is now logged and returned if the app requests array of texture or array of sampler when it is not supported. Please retest and close if fixed.
